core/src/Cabana_Experimental_NeighborList.hpp

```diff
@@ -238,40 +260,60 @@ template <typename DeviceType, typename Slice, typename Tag>
 auto make2DNeighborList( Tag, Slice const &coordinate_slice,
                          typename Slice::size_type first,
                          typename Slice::size_type last,
-                         typename Slice::value_type radius )
+                         typename Slice::value_type radius,
+                         int buffer_size = 0 )
 {
     using MemorySpace = typename DeviceType::memory_space;
     using ExecutionSpace = typename DeviceType::execution_space;
     ExecutionSpace space{};
 
     ArborX::BVH<MemorySpace> bvh( space, coordinate_slice );
 
-    Kokkos::View<int *, DeviceType> indices( "indices", 0 );
-    Kokkos::View<int *, DeviceType> offset( "offset", 0 );
-
     auto const predicates =
         Impl::makePredicates( coordinate_slice, first, last, radius );
 
     auto const n_queries =
         ArborX::AccessTraits<decltype( predicates ),
                              ArborX::PredicatesTag>::size( predicates );
 
+    Kokkos::View<int **, DeviceType> neighbors;
     Kokkos::View<int *, DeviceType> counts( "counts", n_queries );
-    bvh.query(
-        space, predicates,
-        Impl::NeighborDiscriminatorCallback2D_FirstPass<decltype( counts ),
-                                                        Tag>{counts},
-        indices, offset );
+    if ( buffer_size > 0 )
+    {
+        neighbors = Kokkos::View<int **, DeviceType>( "neighbors", n_queries,
+                                                      buffer_size );
+        bvh.query(
+            space, predicates,
+            Impl::NeighborDiscriminatorCallback2D_FirstPass_BufferOptimization<
+                decltype( counts ), decltype( neighbors ), Tag>{counts,
+                                                                neighbors} );
+    }
+    else
```

Comment:
So you don't allow negative buffer here? Maybe at least check?

Reply:
Sure will assert it is > 0